File: frappe_test/frappe_test/report/service_booking_report/service_booking_report.js
// Copyright (c) 2025, Tahsin Ahmed Refat and contributors
// For license information, please see license.txt

frappe.query_reports["Service Booking Report"] = {
	"filters": [
		{
			label: "From Date",
			fieldname: "from_date",
			fieldtype: "Date",
			default: get_first_day_of_month(),
		},
		{
			label: "To Date",
			fieldname: "to_date",
			fieldtype: "Date",
			default: get_last_day_of_month(),
		},
		{
			label: "Customer",
			fieldname: "customer",
			fieldtype: "Link",
			options: "Employee",
			only_select: true,
		},
		{
			label: "Service Type",
			fieldname: "SERVICE_TYPE",
			fieldtype: "Select",
			options: [
				" ",
				"Therapy",
				"Spa",
				"Others"
			],
		},
		{
			label: "Status",
			fieldname: "status",
			fieldtype: "Select",
			options: [
				" ",
				"Requested",
				"Approved",
				"Completed"
			],
		},
	]
};

function get_first_day_of_month() {
  let today = new Date();
  return `${today.getFullYear()}-${("0" + (today.getMonth() + 1)).slice(
    -2
  )}-01`;
}

// Helper function to get the last day of the current month
function get_last_day_of_month() {
  let today = new Date();
  let lastDay = new Date(today.getFullYear(), today.getMonth() + 1, 0); // last day of current month
  return `${lastDay.getFullYear()}-${("0" + (lastDay.getMonth() + 1)).slice(
    -2
  )}-${("0" + lastDay.getDate()).slice(-2)}`;
}

File: frappe_test/frappe_test/report/service_booking_report/service_booking_report.py
# Copyright (c) 2025, Tahsin Ahmed Refat and contributors
# For license information, please see license.txt

import frappe
from frappe.query_builder import DocType
from pypika import Order

def execute(filters=None):
    if not filters:
        filters = {}

    # Columns
    columns = [
        {"label": "Appointment ID", "fieldname": "name", "fieldtype": "Link", "options": "Service Booking", "width": 200},
        {"label": "Customer", "fieldname": "customer", "fieldtype": "Link", "options": "Employee", "width": 200},
        {"label": "Customer Name", "fieldname": "customer_name", "fieldtype": "Data", "width": 350},
        {"label": "Service Type", "fieldname": "service_type", "fieldtype": "Data", "width": 120},
        {"label": "Status", "fieldname": "status", "fieldtype": "Data", "width": 100},
        {"label": "Date", "fieldname": "datetime", "fieldtype": "Datetime", "width": 245},
    ]

    SB = DocType("Service Booking")
    query = (
        frappe.qb.from_(SB)
        .select(
            SB.name,
            SB.customer,
            SB.customer_name,
            SB.service_type,
            SB.status,
            SB.datetime,
        )
    )

    # Dynamically apply filters based on the filters are emtpy or not
    if filters.get("from_date"):
        query = query.where(SB.datetime >= filters["from_date"])
    if filters.get("to_date"):
        query = query.where(SB.datetime <= filters["to_date"])
    if filters.get("customer"):
        query = query.where(SB.customer == filters["customer"])
    if filters.get("SERVICE_TYPE"):
        query = query.where(SB.service_type == filters["SERVICE_TYPE"])
    if filters.get("status"):
        query = query.where(SB.status == filters["status"])

    # Optional: Order by date descending
    query = query.orderby(SB.datetime, order=Order.desc)

    data = query.run(as_dict=True)

    return columns, data
